index.js

```diff
@@ -2109,17 +2137,24 @@
                     String(t)));
   }
 
-  //  showValuesAndTypes :: (Array Type, Array Any, Integer) -> String
-  function showValuesAndTypes(env, values, pos) {
+  //  showValuesAndTypes :: ... -> String
+  function showValuesAndTypes(
+    env,            // :: Array Type
+    typeInfo,       // :: TypeInfo
+    values,         // :: Array Any
+    pos             // :: Integer
+  ) {
     return String(pos) + ')  ' + Z.map(function(x) {
       var types = determineActualTypesLoose(env, [x]);
-      return Z.toString(x) + ' :: ' + Z.map(showType, types).join(', ');
+      return Z.toString(x) +
+             ' :: ' +
+             Z.map(showType(typeInfo), types).join(', ');
```

Comment:
You could assign the result of `showType(typeInfo)` to a variable to make better use of the optimizations proposed before.